Userland/Services/FileSystemAccessServer/ClientConnection.cpp

```diff
@@ -0,0 +1,166 @@
+
+        if (requested_access & O_WRONLY)
+            builder.append('w');
+
+        if (requested_access & O_CREAT)
+            builder.append('c');
+
+        auto access_string = builder.to_string();
+
+        auto result = GUI::MessageBox::show(nullptr, String::formatted("Give {} \"{}\" access to \"{}\"?", sender, access_string, path), "File Permissions Requested", GUI::MessageBox::Type::Warning, GUI::MessageBox::InputType::YesNo);
+
+        approved = result == GUI::MessageBox::ExecYes;
+
+        if (approved) {
+            // Intentionally overwrite existing access
+            m_approved_files.set(path, requested_access);
+        }
+    }
+
+    if (approved) {
+        auto fd = open(path.characters(), requested_access);
+
+        if (fd < 0) {
+            dbgln(String::formatted("FileSystemAccessServer: Couldn't open {}, error {}", path, OSError(errno).string()));
+
+            return { errno, Optional<IPC::File> {} };
+        }
+
+        return { 0, IPC::File(fd, IPC::File::CloseAfterSending) };
+    }
+
+    return { -1, Optional<IPC::File> {} };
+}
+
+Messages::FileSystemAccessServer::PromptOpenFileResponse ClientConnection::prompt_open_file(String const& path_to_view, int requested_access)
+{
+    auto main_window = GUI::Window::construct();
+    Optional<String> user_picked_file = GUI::FilePicker::get_open_filepath(main_window, "Select file", path_to_view);
+
+    if (user_picked_file.has_value()) {
+        auto fd = open(user_picked_file->characters(), requested_access, 0644);
+
+        if (fd < 0) {
+            dbgln(String::formatted("FileSystemAccessServer: Couldn't open {}, error {}", user_picked_file.value(), OSError(errno).string()));
+
+            return { errno, Optional<IPC::File> {}, Optional<String> {} };
+        }
+
+        return { 0, IPC::File(fd, IPC::File::CloseAfterSending), user_picked_file.value() };
+    }
+
+    return { -1, Optional<IPC::File> {}, Optional<String> {} };
+}
+
+Messages::FileSystemAccessServer::PromptSaveFileResponse ClientConnection::prompt_save_file(String const& name, String const& ext, String const& path_to_view, int requested_access)
+{
+    auto main_window = GUI::Window::construct();
+    Optional<String> user_picked_file = GUI::FilePicker::get_save_filepath(main_window, name, ext, path_to_view);
+
+    if (user_picked_file.has_value()) {
+        auto fd = open(user_picked_file->characters(), requested_access, 0644);
+
+        if (fd < 0) {
+            dbgln(String::formatted("FileSystemAccessServer: Couldn't open {}, error {}", user_picked_file.value(), OSError(errno).string()));
+
+            return { errno, Optional<IPC::File> {}, Optional<String> {} };
+        }
+
+        return { 0, IPC::File(fd, IPC::File::CloseAfterSending), user_picked_file.value() };
+    }
+
+    return { -1, Optional<IPC::File> {}, Optional<String> {} };
+}
+
+void ClientConnection::request_download_file(String const& sender, URL const& url, String const& path)
```

Comment:
What if the file already exists?

Reply:
I've changed the MessageBox text for when it's overwriting an existing file.

I can see `request_download_file` changing a lot when it starts to get used but I don't have a place to use it right now.
I think it might be best to delete it until a use case appears, how does that sound?

Reply:
Moved all download stuff to a separate branch, I'll start a separate PR for it later.